File: web-client/src/app/Models/Review.ts
import { IDiningHall, DiningHall } from './DiningHall';
import { Dorm, ILargeDorm } from './Dorm';
import { IParkingGarage, ParkingGarage } from './ParkingGarage';
import { ICourse, Course } from './Course';
import { ReviewSummary } from './ReviewSummary';
import { IUser } from './User';

export interface IReview {
  id: number;
  reviewText: string;
  starRating: number;
  timeCreated: Date;
  averageVote: number;
  userId: number;
  dormId: number;
  user: IUser;
  dorm?: ILargeDorm;
  parkingGarage?: IParkingGarage;
  userVoteType: UserVoteType;
  diningHall?: IDiningHall;
  course?: ICourse;
}
export enum UserVoteType {
  UserUpvoted,
  UserDownvoted,
  UserNeutral,
}
export type Review = IReview;

export type Reviewable = Dorm | ParkingGarage | DiningHall | Course;

export class SaveReview {
  constructor(props: {
    reviewText: string;
    rating: string;
    userId: number;
    parkingGarageId?: Number;
    dormId?: Number;
    diningHallId?: Number;
    courseId?: Number;
  }) {
    Object.assign(this, props);
  }
}

export type SaveReviewResponse = {
  review: Review;
  summary: ReviewSummary;
};

File: api/Models/Review.cs
namespace api.Models;

using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;
using Microsoft.EntityFrameworkCore;

#nullable enable

public class Review
{
    [Key]
    public int Id { get; set; }
    public required string ReviewText { get; set; }

    [Precision(2, 1)]
    public decimal StarRating { get; set; }
    public DateTime TimeCreated { get; set; }

    [ForeignKey("User")]
    public int UserId { get; set; }

    [JsonIgnore]
    public virtual User User { get; set; }

    [ForeignKey("Dorm")]
    public int? DormId { get; set; }

    [JsonIgnore]
    public virtual Dorm? Dorm { get; set; }

    [ForeignKey("ParkingGarage")]
    public int? ParkingGarageId { get; set; }

    [JsonIgnore]
    public virtual ParkingGarage? ParkingGarage { get; set; }

    [ForeignKey("DiningHall")]
    public int? DiningHallId { get; set; }

    [JsonIgnore]
    public virtual DiningHall? DiningHall { get; set; }

    [ForeignKey("Course")]
    public int? CourseId { get; set; }

    [JsonIgnore]
    public virtual Course? Course { get; set; }
}
